src/Microsoft.OData.Core/Json/ODataJsonTextWriter.cs

```diff
@@ -299,54 +299,50 @@ public override Task FlushAsync()
         }
 
         /// <inheritdoc/>
-        public override Task WriteAsync(char value)
+        public override async Task WriteAsync(char value)
         {
-            return TaskUtils.GetTaskForSynchronousOperation(() =>
-                this.Write(value));
+            await this.WriteEscapedCharValueAsync(value).ConfigureAwait(false);

```

Comment:
Is it necessary to async/await this given that the method being called already returns an async task? Or it doesn't make a difference? I assume we incur more overhead in this case compared to just returning the task? I could be wrong.

Reply:
According to the discussion here: https://stackoverflow.com/questions/38017016/async-task-then-await-task-vs-task-then-return-task the differences lie in exception handling and the state machine generated for async/await by the compiler. The discussions also suggest that there are optimizations to prevent unnecessary awaiting. But according to [the linked sharplab code sample](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQFEAeAHJzkCWA9gHYBqUCBUwANigD4ACADAARMCMA3ALABQrDpwCsffgKYBmDgCY2AYTYBvAW3UcZTABwcAbGwCCyAJ4kAxoYDuUAjAAUASjYBeAHwcAnEdMWFUWnoIJ3ENNjUNaX0fM3MAJQh4BBInVw9jWP9AiGDHUI0I9VwCADdYCGiMvwCg1Pd9ADoFIgBbbHoYCAATJj1xAF8gA) it appears that the state machine instance is generated regardless. I think this could lead to unnecessary allocations.

Reply:
My thoughts are that we don't need async here.
But am not sure if there is extra overhead by having async

Reply:
@habbes @KenitoInc Eliding `async` and `await` can lead to unexpected pitfalls/behaviour. Reference to [this](https://blog.stephencleary.com/2016/12/eliding-async-await.html) article plus [Stephen Toub's  article on the cost of `async` and `await`](https://docs.microsoft.com/en-us/archive/msdn-magazine/2011/october/asynchronous-programming-async-performance-understanding-the-costs-of-async-and-await), the state machine for `async` methods will capture exceptions from your code and place them on the returned task. Without the `async` keyword, the exception is raised directly rather than going on the task:
```csharp
public async Task<string> GetWithKeywordsAsync()
{
    string url = /* Something that can throw an exception */;
    return await DownloadStringAsync(url);
}

public Task<string> GetElidingKeywordsAsync()
{
    string url = /* Something that can throw an exception */;
    return DownloadStringAsync(url);
}
```
These methods work exactly the same as long as the calling method does something like this:
```csharp
var result = await GetWithKeywordsAsync();

var result = await GetElidingKeywordsAsync();
```
However, if the method call is separated from the `await`, then the semantics are different:
```csharp
var task = GetWithKeywordsAsync();
var result = await task; // Exception thrown here

var task = GetElidingKeywordsAsync(); // Exception thrown here
var result = await task;
```
Eliding the keywords in this case causes different (and unexpected) exception behavior.

Honestly I'd rather focus on micro optimizations separately after a analyzing the costs carefully plus first determining whether the compiler does those kind of optimizations by itself.
Plus I'd hope to receive views on the same from other reviewers before embarking on those kind of manual interventions.

Reply:
But in his particular case we are not doing anything that could throw an exception before the await. We're simply fowarding the arguments as they are to an async call, with no intermerdiary checks or preprocessing calls. Also, since this that's called by another async method with will be called using async/await, any exception that it would throw would be caught by the async method that calls it, so it wouldn't end being unhandled. Aside from the optimization concern, I was just wondering why it was necessary to have the await there in the first place.

But if you feel like it's risky to remove it, I have not objections to keeping it there.